codice/ddf
# DDF-4133 Improve communication with Solr

#### What does this PR do?
The two goals of this PR  are:
    1. Use SSL to communicate between DDF and Solr with no client authorization. 
    2. Use basic auth over SSL so that Solr can grant access to DDF and no one else.

#### Who is reviewing it? 
@mcalcote @austinsteffes @paouelle 

#### Ask 2 committers to review/merge the PR and tag them here.
@clockard
@stustison 
@pklinef 

#### Tag the appropriate codice teams
@codice/solr 
@codice/security 

#### How should this be tested?
For heroing, there are several scenarios to test:
 1. Verify DDF "just works" out of the box. This is, can it communicate with Solr without configuration.
 2.  Attempt to log into the Solr admin console. Verify that it does not permit access unless provided with the username and password.
 3.  Change the password and username for the Solr user. Update the DDF system.properties file with the new username and password. Restart DDF and Solr. Verify they can still communicate.

#### Any background context you want to provide?
Getting this to work was a little tricky because the Apache HTTP client only sends basic auth headers for GET requests. However, Solr requires that the basic auth header be present for POST and DELETE requests as well. The effort is large enough that some work must exist as sub-tickets.

Documentation - https://codice.atlassian.net/browse/DDF-4148
Security - https://codice.atlassian.net/browse/DDF-4149 

#### What are the relevant tickets?
[](https://codice.atlassian.net/browse/)
#### Screenshots

#### Checklist:
- [N/A] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
